src/Microsoft.AspNetCore.Server.IISIntegration/Server/IO/AsyncWriteOperationBase.cs

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using Microsoft.AspNetCore.HttpSys.Internal;
+
+namespace Microsoft.AspNetCore.Server.IISIntegration
+{
+    internal abstract class AsyncWriteOperationBase : AsyncIOOperation
+    {
+        private const int HttpDataChunkStackLimit = 128; // 16 bytes per HTTP_DATA_CHUNK
+
+        private IntPtr _requestHandler;
+        private ReadOnlySequence<byte> _buffer;
+        private MemoryHandle[] _handles;
+
+        public void Initialize(IntPtr requestHandler, ReadOnlySequence<byte> buffer)
+        {
+            _requestHandler = requestHandler;
+            _buffer = buffer;
+        }
+
+        protected override unsafe bool InvokeOperation(out int hr, out int bytes)
+        {
+            if (_buffer.Length > int.MaxValue)
+            {
+                throw new InvalidOperationException($"Writes larger then {int.MaxValue} are not supported.");
+            }
+
+            bool completionExpected;
+            var chunkCount = GetChunkCount();
+
+            var bufferLength = (int)_buffer.Length;
+
+            if (chunkCount < HttpDataChunkStackLimit)
+            {
+                // To avoid stackoverflows, we will only stackalloc if the write size is less than the StackChunkLimit
+                // The stack size is IIS is by default 128/256 KB, so we are generous with this threshold.
+                var chunks = stackalloc HttpApiTypes.HTTP_DATA_CHUNK[chunkCount];
+                hr = WriteSequence(chunkCount, _buffer, chunks, out completionExpected);
+            }
+            else
+            {
+                // Otherwise allocate the chunks on the heap.
+                var chunks = new HttpApiTypes.HTTP_DATA_CHUNK[chunkCount];
+                fixed (HttpApiTypes.HTTP_DATA_CHUNK* pDataChunks = chunks)
+                {
+                    hr = WriteSequence(chunkCount, _buffer, pDataChunks, out completionExpected);
+                }
+            }
+
+            bytes = bufferLength;
+            return !completionExpected;
+        }
+
+        public override void FreeOperationResources(int hr, int bytes)
+        {
+            // Free the handles
+            foreach (var handle in _handles)
+            {
+                handle.Dispose();
```

Comment:
Based on the above feedback, we should look into freeing the operation resources after calling WriteChunks rather than when the operation returns.

Reply:
This one is for pinning pipe blocks.